### PR TITLE
Refinement Operator in FiniteElementSpace [refine-operator-dev]

### LINKED ISSUE #247: Reduce memory for Mesh refinement, especially for high-order meshes
Refining a high-order `Mesh` uses a lot more memory than expected. This seems to be due to the use of a fully assembled matrix to interpolate the node coordinates.

### PR DESCRIPTION
**Pull request by jakubcerveny**
This PR addresses issue #247 by replacing the function RefinementMatrix in FiniteElementSpace with the class RefinementOperator. This means the refinement matrix is no longer constructed explicitly, to save both time and RAM (lots of RAM). There should be no explicit effect of this outside of FiniteElementSpace, everything should work as before.

I left some testing code in ex1, ex1p, ex2, ex3, ex4 that will not be merged. At the end of these examples the solution is refined and interpolated several times. 

The RAM savings should be substantial, for example, the modified ex1p won't finish with MFEM master on my machine with 32 GB RAM, with this commandline:

    mpirun -np 4 ex1p -m ../data/ball-nurbs.mesh -o 4

TODO:
- [x] Update `CHANGELOG`
- [x] Fix `ex3p` failures
- [x] Revert testing changes in the examples

Resolves #247

### CODE REVIEW

File: fem/fespace.hpp [hunk 1]
```diff
@@ -118,11 +122,22 @@ class FiniteElementSpace
...
+   /// GridFunction interpolation operator applicable after mesh refinement.
+   class RefinementOperator : public Operator
+   {
+      const FiniteElementSpace* fespace;
+      DenseTensor localP;
+      Table* old_elem_dof; // owned
+
+   public:
+      RefinementOperator(const FiniteElementSpace* fespace,
+                         Table *old_elem_dof/*takes ownership*/, int old_ndofs);
```
Comment (tzanio) [inline fem/fespace.hpp]: Document that in a Doxygen comment for the method

**Comment by tzanio** [comment]
Thanks Jakub! Approved modulus TODO items in the PR description.

**Comment by tzanio** [comment]
The inability to refine after `ReorientTetMesh` is a feature :), see

https://github.com/mfem/mfem/blob/637ea4142d231f154b057ec219dae5835fc19e54/mesh/mesh.hpp#L883

**Comment by v-dobrev** [comment]
This restriction (needed for Nedelec elements on tets) can be removed using a suitable prolongation operator to match the dofs on neighbor triangular faces with arbitrary orientations. The mapping there is not just a permutation, that's why we need the more general prolongation.

**Comment by jakubcerveny** [comment]
OK, I can put back the full assembly option. I'm worried it will clutter the interface though. How do you imagine the user will control this?

**Comment by v-dobrev** [comment]
One approach we can use is similar to class ParBilinearForm: add a method that takes an Operator::Type and store that in an OperatorHandle that will hold the type and, later, the constructed operator. For the SparseMatrix case we can use the existing Operator:: MFEM_SPARSEMAT. For the other case, we'll need to add a new type, e.g. Operator::AUTO, or Operator::GENERIC which will leave the choice to the library.

What do you think?

**Comment by v-dobrev** [comment]
One question with my proposal above is: since in class (Par)FiniteElementSpace, we have various types of transformation operators (refine, de-refine, rebalance), do we want to have a separate OperatorHandle for each or just use one and let the user switch the type based on the last transformation? Using only one OperatorHandle will keep the interface simpler, so it is probably the better choice.

**Comment by jakubcerveny** [comment]
RefinementMatrix is back. I studied how OperatorHandle is used in the code (this is new for me) but in the end implemented a trivial solution without it. I just didn't feel it's the perfect fit for this (adding Operator::GENERIC felt kind of artificial; getting rid of own_T would be nice but then OperatorHandle doesn't really work as a shared pointer so I don't know). But I may be wrong. Feel free to modify this to your liking, @v-dobrev

File: CHANGELOG [hunk 2]
```diff
@@ -52,6 +52,10 @@ Miscellaneous
 
 - Add support for building a shared version of the MFEM library with GNU make.
 
+- In FiniteElementSpace, the fully assembled refinement matrix is now replaced
```
Comment (tzanio) [inline CHANGELOG]: Revise, e.g. "... is now replaced by default by... . The fully assembled matrix is still available if ..."

**Comment by tzanio** [comment]
Thanks @jakubcerveny!

Can you also merge `master` and resolve the conflicts?

I am not a huge fan of `FESpace::PreferActionOnlyUpdateOperator(bool`... how about `FESpace::AssembleUpdateMatrix(bool)` instead (with the opposite meaning)?

**Comment by jakubcerveny** [comment]
`AssembleUpdateMatrix` does sound better but it also looks like a method to actually assemble something. I think it should be clear that the method is only setting some option/preference (but I agree `PreferActionOnlyUpdateOperator` is not great).

**Comment by v-dobrev** [comment]
I implemented my approach with `OperatorHandle` in the branch `oper-type-any` and made a for it (PR #418) into this branch. Take a look, and if you like it (or you are just OK with it) - approve the PR.

**Comment by jakubcerveny** [comment]
Thanks Veselin. I think this is ready now.

**Comment by v-dobrev** [comment]
@tzanio, did you take a look at, and are you OK with, the changes from #418?

**Comment by tzanio** [comment]
> @tzanio, did you take a look at, and are you OK with, the changes from #418?

yes and yes

**Pull request merged**